engine/access/ingestion/engine_test.go

```diff
@@ -257,8 +271,86 @@ func (s *Suite) TestOnFinalizedBlock() {
 }
 
+// TestOnFinalizedBlockSeveralBlocksAhead checks OnFinalizedBlock with a block several blocks newer than the last block processed
+func (s *Suite) TestOnFinalizedBlockSeveralBlocksAhead() {
+	cluster := new(protocol.Cluster)
+	epoch := new(protocol.Epoch)
+	epochs := new(protocol.EpochQuery)
+	snap := new(protocol.Snapshot)
+
+	epoch.On("ClusterByChainID", mock.Anything).Return(cluster, nil)
+	epochs.On("Current").Return(epoch)
+	snap.On("Epochs").Return(epochs)
+
+	// prepare cluster committee members
+	clusterCommittee := unittest.IdentityListFixture(32 * 4).Filter(filter.HasRole[flow.Identity](flow.RoleCollection)).ToSkeleton()
+	cluster.On("Members").Return(clusterCommittee, nil)
+
+	lastFinalizedHeight := s.finalizedBlock.Height
+	s.blocks.On("GetLastFullBlockHeight").Return(lastFinalizedHeight, nil).Maybe()
+
+	irrecoverableCtx, _ := irrecoverable.WithSignaler(s.ctx)
+	eng := s.initIngestionEngine(irrecoverableCtx)
+
+	blkCnt := 3
+	startHeight := s.finalizedBlock.Height + 1
+	blocks := make([]flow.Block, blkCnt)
+
+	// generate the test blocks, cgs and collections
+	for i := 0; i < blkCnt; i++ {
+		block := s.generateBlock(clusterCommittee, snap)
+		block.Header.Height = startHeight + uint64(i)
+		s.blockMap[block.Header.Height] = &block
+		blocks[i] = block
+		s.mockCollectionsForBlock(block)
+		s.finalizedBlock = block.Header
+	}
+
+	// block several blocks newer than the last block processed
+	hotstuffBlock := hotmodel.Block{
+		BlockID: blocks[2].ID(),
+	}
+	// for each of the guarantees, we should request the corresponding collection once
+	needed := make(map[flow.Identifier]struct{})
+	for _, guarantee := range blocks[0].Payload.Guarantees {
+		needed[guarantee.ID()] = struct{}{}
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(4)
+
+	s.request.On("EntityByID", mock.Anything, mock.Anything).Run(
+		func(args mock.Arguments) {
+			collID := args.Get(0).(flow.Identifier)
+			_, pending := needed[collID]
+			s.Assert().True(pending, "collection should be pending (%x)", collID)
+			delete(needed, collID)
+			wg.Done()
+		},
+	)
+
+	// expected next block after last block processed
+	s.blocks.On("IndexBlockForCollections", blocks[0].ID(), []flow.Identifier(flow.GetIDs(blocks[0].Payload.Guarantees))).Return(nil).Once()
+	s.results.On("Index", mock.Anything, mock.Anything).Return(nil)
+
+	eng.OnFinalizedBlock(&hotstuffBlock)
+
+	s.Assertions.Eventually(func() bool {
+		wg.Wait()
+		return true
+	}, time.Millisecond*20, time.Millisecond)
+
+	s.headers.AssertExpectations(s.T())
+	s.blocks.AssertNumberOfCalls(s.T(), "IndexBlockForCollections", 1)
+	s.request.AssertNumberOfCalls(s.T(), "EntityByID", len(blocks[0].Payload.Guarantees))
+	s.request.AssertNumberOfCalls(s.T(), "Index", len(blocks[0].Payload.Seals))
+}
+
 // TestOnCollection checks that when a Collection is received, it is persisted
 func (s *Suite) TestOnCollection() {
+	irrecoverableCtx, _ := irrecoverable.WithSignaler(s.ctx)
```

Comment:
consider switching these over to mocks:
https://github.com/onflow/flow-go/blob/c6e460dbef0fcb6a25c2ef88a325dbc93caaf763/module/irrecoverable/unittest.go#L14

this will gracefully handle failures if unexpected errors are thrown

Reply:
[ad41a50](https://github.com/onflow/flow-go/pull/5714/commits/ad41a50b4df8d48a7bd1d6ea4da60f48df1b1934)